### DISCUSSION TITLE
Static build problems again

### DISCUSSION THREAD

**Discussion by speedgeil** [ImageMagick/ImageMagick#5203] [Help]
I'm on MacOS but Linux should be quite similar.
I managed to build a static version of Imagemagick but i'm having problems with RAW support.

As far as i understand Imagemagick uses ufraw-batch on Linux and Mac while it uses dcraw on windows.
My problem is that i cant find a source for ufraw-batch. I tried with macports ufraw but i can't get it to compile static. Can someone point me on a source or alternatively can i use libraw or dcraw?

Basically i only need a static build with raw and tif support.

**Comment by snibgo** [comment]
The usual build on Windows is with libraw.

dcraw is an external program. You can create an entry in delegates.xml like this:
```
  <delegate decode="dng:decode" stealth="True" command="dcraw.exe -6 -W -O &quot;%u.ppm&quot; &quot;%i&quot;"/>
```

**Reply by speedgeil** [reply]
Thank you, on windows i have everything i need because there is a ready made static version. My problem relates to MacOS. 
It's quite hard to wrap my head around this as the IM user guide is not really a help and i don't know what to google/where to start. I'm also a windows guy and MacOS / XCode give me the headaches of my life.

**Comment by snibgo** [comment]
What happens when you use IM to read a raw file? What are the eror messages?

On Mac, you have built a static version of IM. Correct? Please show the text output from "magick -version". That will say whether your IM has libraw.

Do you have ufraw-batch or dcraw on your Mac? If not, you will need to install one of them. The binary code.

When you have ufraw-batch or dcraw, you can insert a line into delegates.xml.

**Reply by speedgeil** [reply]
Obviously i have betrayed myself and what i have compiled is not static. Once i removed the files from /usr/local/bin, the IM binaries in my app's folder stopped working (in fact they never worked). So i'll have to figure out why and what first.

However, this is what magick -version gives me: 

Version: ImageMagick 7.1.0-38 beta Q16 x86_64 1b8963a40:20220605 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC 
Delegates (built-in): bzlib fontconfig freetype gvc heic jbig jng jp2 jpeg lcms lzma openexr pangocairo png raw tiff webp xml zlib
Compiler: gcc (4.2)

**Comment by snibgo** [comment]
> Delegates (built-in): ... raw ...

So, IM has been built with libraw, and I expect that IM can read read raw files. Can it? Does it throw any errors?

**Reply by speedgeil** [reply]
It actually can but it's not static (portable). Once i copy IM to another computer it will stop working.

There are some things i don't understand when using ./configure. And compiling libraw for being portable is another problem. I'll try figuring that out first.

**Reply by speedgeil** [reply]
So... i just don't get it. 
Imagemagick static for Windows comes with binaries for dcraw, libraw and everything. I just want the same on MacOS. No dependencies outside the imagemagick folder. Everything should be in that folder. I want it on an USB stick, plug it into another computer and then it should run. Is that possible at all on Linux/MacOS?
The imagemagick user guide completely fails at explaining static builds. It's even missing own configure flags like "--with-raw".
For example I configured libraw, tiff, jpeg, png and lcms with --disable-shared like explained by magick here: https://legacy.imagemagick.org/discourse-server/viewtopic.php?t=11137
When running the configure script in the IM root folder it will recognize jpeg and tiff but not png, lcms and not raw. According to the log it doesn't find them 

`-------------------------------------------------------------
checking for libpng >= 1.0.0... no

-------------------------------------------------------------
checking for libraw_r >= 0.14.8... no

-------------------------------------------------------------
checking for TIFF... 
checking for tiff.h... yes
checking for tiffio.h... yes
checking for TIFFOpen in -ltiff... yes
checking for TIFFClientOpen in -ltiff... yes
checking for TIFFIsByteSwapped in -ltiff... yes
checking for TIFFReadRGBATile in -ltiff... yes
checking for TIFFReadRGBAStrip in -ltiff... yes
checking if TIFF package is complete... yes
checking for tiffconf.h... yes
checking for TIFFIsCODECConfigured... yes
checking for TIFFMergeFieldInfo... yes
checking for TIFFIsBigEndian... yes
checking for TIFFSetErrorHandlerExt... yes
checking for TIFFSetTagExtender... yes
checking for TIFFReadEXIFDirectory... yes
checking for TIFFReadGPSDirectory... no
checking for TIFFSetWarningHandlerExt... yes
checking for TIFFSwabArrayOfTriples... yes`

Another thing i don't get is that the delegates.xml states ufraw as dng coder. Why does everyone say libraw?
And then there is the "MacPorts method". Means i would install needed delegates with MacPorts. That's cool and it will allow me to run imagemagick "statically" on one computer? That's maybe static, but not portable right? Or is there a way to cobble the delegates and dependencies into the IM folder while installing?

I admit, i have limited knowledge regarding compiling. And yes, this is a Linux/clang/ whatever topic. But really, there is literally no docs for compiling static. Neither on Linux nor MacOs. Homebrew was a solution for Mac users to compile static as far as i read earlier. Now Homebrew won't allow brewing with options  any longer since two years and Mac Users are totally left alone? It would be nice to have at least some instructions for a full static build. Honestly i find that is an issue as 70% of the media world is still working on Macs.

As said, limited knowledge. But it's really hard gaining knowledge on these topics.

**Comment by bioworldmerch** [comment]
Not sure if it will help but here's how I got my static install working: 

https://github.com/ImageMagick/ImageMagick/discussions/5215
